### PR TITLE
allow to generate imports w/o using dynamic expressions

### PR DESCRIPTION
**Pull request by mmomtchev**
Refs: #20503

* Render the code fragments referencing `Module['mainScriptUrlOrBlob']` and `Module['locateFile']` optional
* Add `INCOMING_MODULE_JS_API` processing logic to `preprocessor.js`

### CODE REVIEW

File: src/library_pthread.js [hunk 1]
```diff
@@ -372,7 +372,12 @@ var LibraryPThread = {
         // it could load up the same file. In that case, developer must either deliver the Blob
         // object in Module['mainScriptUrlOrBlob'], or a URL to it, so that pthread Workers can
         // independently load up the same main application file.
-        'urlOrBlob': Module['mainScriptUrlOrBlob']
+        'urlOrBlob':
+#if expectToReceiveOnModule('mainScriptUrlOrBlob')
+        Module['mainScriptUrlOrBlob']
+#else
+        undefined
+#endif
```
Comment (sbc100) [inline src/library_pthread.js:380]: Can we just skip the `urlOrBlob` when we don't expect to receive `mainScriptUrlOrBlob`.  i.e. can you wrap the member declaration along with the whole comment above in he `if/endif`?

File: test/test_other.py [hunk 2]
```diff
@@ -356,6 +356,21 @@ def test_emcc_output_mjs_web_no_import_meta(self):
     self.assertNotContained('import.meta.url', src)
     self.assertContained('export default Module;', src)
 
+  def test_emcc_output_mjs_no_import_expression(self):
```
Comment (sbc100) [inline test/test_other.py:359]: How about `test_mjs_no_dynamic_import`?

File: tools/preprocessor.js [hunk 3]
```diff
@@ -55,6 +55,12 @@ const inputFile = arguments_[1];
 const expandMacros = arguments_.includes('--expandMacros');
 
 load(settingsFile);
+
+if (!ALL_INCOMING_MODULE_JS_API.length) {
+  ALL_INCOMING_MODULE_JS_API = INCOMING_MODULE_JS_API;
+}
+INCOMING_MODULE_JS_API = new Set(INCOMING_MODULE_JS_API);
```
Comment (sbc100) [inline tools/preprocessor.js:62]: Why is this part needed?

File: test/test_other.py [hunk 4]
```diff
@@ -356,6 +356,21 @@ def test_emcc_output_mjs_web_no_import_meta(self):
...
+                      '-pthread',
+                      '-sMODULARIZE=1',
+                      '-sEXPORT_ES6',
+                      '-sENVIRONMENT=web,worker',
+                      '-sUSE_ES6_IMPORT_META=1',
+                      '-sINCOMING_MODULE_JS_API=[]'
+                    ])
+    src = read_file('hello_world.mjs')
+    self.assertNotContained('import(e.data.urlOrBlob)', src)
+    self.assertContained('import.meta.url', src)
```
Comment (RReverser) [inline test/test_other.py:372]: I'd test that a static `import('./...')` is present here.

File: tools/preprocessor.js (hunk 3 above)
Comment (mmomtchev) [reply tools/preprocessor.js:62]: Because of this line: https://github.com/emscripten-core/emscripten/blob/fcb33368d64873786762df2dd44cc9a18d4b1286/src/parseTools.js#L752

It expects a `Set`. `compiler.js` does the same.

Comment (kripken) [reply tools/preprocessor.js:62]: It looks like `compiler.js` turns both `INCOMING_MODULE_JS_API` and `ALL_INCOMING_MODULE_JS_API` into sets:

https://github.com/emscripten-core/emscripten/blob/a7d1f86b121e4223787db99ff9134ad964b08ddf/src/compiler.js#L76-L77

But this code only does that for the former. Why is there a difference?

**Review by kripken — COMMENTED** [review]
Another approach that might be worth considering is, instead of adding `expectToReceiveOnModule('locateFile')` checks in various places, to do this up front in one place:
```
#if expectToReceiveOnModule('locateFile')
var locateFile = Module['locateFile'];
#else
var locateFile = (file) => file; // no-op
#endif
```
Then all the code can assume `locateFile` already exists, everywhere. With closure compiler the no-op function will get inlined and optimized away. Thoughts?

File: tools/preprocessor.js (hunk 3 above)
Comment (mmomtchev) [reply tools/preprocessor.js:62]: Since I used only `INCOMING_MODULE_JS_API` I did not add the rest.

Comment (sbc100) [reply tools/preprocessor.js:62]: But I'm curious why this change required this bug fix?  Does it fail without this change?

Comment (mmomtchev) [reply tools/preprocessor.js:62]: Yes, an array does not have `.has()`, only `Set` does

**Comment by mmomtchev** [comment]
Maybe you should start by launching the unit tests? Normally, your time should be more expensive than that of the Github runner?

File: src/library_pthread.js [hunk 5]
```diff
@@ -367,16 +367,19 @@ var LibraryPThread = {
...
         // to access the URL of the current script that could be passed to a Web Worker so that
         // it could load up the same file. In that case, developer must either deliver the Blob
         // object in Module['mainScriptUrlOrBlob'], or a URL to it, so that pthread Workers can
         // independently load up the same main application file.
-        'urlOrBlob': Module['mainScriptUrlOrBlob']
+        'urlOrBlob':
+        Module['mainScriptUrlOrBlob']
 #if !EXPORT_ES6
         || _scriptDir
 #endif
```
Comment (sbc100) [inline src/library_pthread.js:380]: Do we want to set `urlOrBlob` to `_scriptDir` here?

e.g. in the non EXPORT_ES6 case should this be:
```
#if expectToReceiveOnModule('mainScriptUrlOrBlob')
  'urlOrBlob': Module['mainScriptUrlOrBlob'] || _scriptDir,
#else
  'urlOrBlob': _scriptDir,
#endif
```

Actually it looks like we already have a helper function so you should be able to do:

```
 'urlOrBlob': {{{ makeModuleReceiveExpr('mainScriptUrlOrBlob', '_scriptDir') }}}
```

**Comment by sbc100** [comment]
> Maybe you should start by launching the unit tests? Normally, your time should be more expensive than that of the Github runner?

I'm not sure I understand this comment.  Who is it that you are you suggesting should launch the unit tests?

**Comment by mmomtchev** [comment]
It means that I do not think that there any chances of having a meaningful working relation with someone taking part in an organized crime against me.
